java/memory/src/main/java/org/apache/arrow/memory/util/hash/ArrowBufHasher.java

```diff
@@ -17,93 +17,24 @@
- * It first splits the memory region into small segments with 8 bytes, 4 bytes and 1 byte,
- * and calculates hash codes for them separately. It produces the final hash code by combining
- * the hash codes and finalizing the resulting hash code.
- *
  * <p>
- *   To compute the hash code, the user simply calls the hashCode methods with the starting
- *   address and length of the memory region.
- * </p>
- * <p>
- *   A default light-weight implementation of this class is given in {@link DirectHasher}. However, the users can
- *   devise their own customized hasher by sub-classing this method and overriding the abstract methods.
- *   In particular
- *   <li>
- *     {@link ArrowBufHasher#combineHashCode(int, int)} provides the method for combining hash
- *     codes for individual small segments.
- *   </li>
- *   <li>
- *     {@link ArrowBufHasher#finalizeHashCode(int)} provides the method for finalizing the hash code.
- *   </li>
- *   <li>
- *     {@link ArrowBufHasher#getByteHashCode(byte)} provides the method for calculating the hash code
- *     for 1-byte memory segment.
- *   </li>
- *   <li>
- *     {@link ArrowBufHasher#getIntHashCode(int)} provides the method for calculating the hash code
- *     for 4-byte memory segment.
- *   </li>
- *   <li>
- *     {@link ArrowBufHasher#getLongHashCode(long)} provides the method for calculating the hash code
- *     for 8-byte memory segment.
- *   </li>
+ *   A default light-weight implementation is given in {@link SimpleHasher}.
  * </p>
  */
-public abstract class ArrowBufHasher {
-
-  public static final boolean LITTLE_ENDIAN = ByteOrder.nativeOrder() == ByteOrder.LITTLE_ENDIAN;
+public interface ArrowBufHasher {
 
   /**
    * Calculates the hash code for a memory region.
    * @param address start address of the memory region.
    * @param length length of the memory region.
    * @return the hash code.
    */
-  public int hashCode(long address, int length) {
-    int hashValue = 0;
-    int index = 0;
-    while (index + 8 <= length) {
-      long longValue = getLong(address + index);
-      if (!LITTLE_ENDIAN) {
-        // assume the buffer is in little endian
-        longValue = Long.reverseBytes(longValue);
-      }
-      int longHash = getLongHashCode(longValue);
-      hashValue = combineHashCode(hashValue, longHash);
-      index += 8;
-    }
-
-    while (index + 4 <= length) {
-      int intValue = getInt(address + index);
-      if (!LITTLE_ENDIAN) {
-        intValue = Integer.reverseBytes(intValue);
-      }
-      int intHash = getIntHashCode(intValue);
-      hashValue = combineHashCode(hashValue, intHash);
-      index += 4;
-    }
-
-    while (index < length) {
-      byte byteValue = getByte(address + index);
-      int byteHash = getByteHashCode(byteValue);
-      hashValue = combineHashCode(hashValue, byteHash);
-      index += 1;
-    }
-
-    return finalizeHashCode(hashValue);
-  }
+  int hashCode(long address, int length);
```

Comment:
does this need to be in the API ? it's hard to check for the validity even in debug code.

Reply:
I am not sure either. Maybe occasionally we need to calculate the hash code for an on-heap memory region?